packages/react/select/src/Select.tsx

```diff
@@ -306,34 +293,83 @@ SelectIcon.displayName = ICON_NAME;
 
 const CONTENT_NAME = 'SelectContent';
 
+type ReactOption = React.ReactElement<React.ComponentProps<'option'>>;
+
+type SelectNativeOptionsContextValue = {
+  onNativeOptionAdd(option: ReactOption): void;
+  onNativeOptionRemove(option: ReactOption): void;
+};
+const [SelectNativeOptionsProvider, useSelectNativeOptionsContext] =
+  createSelectContext<SelectNativeOptionsContextValue>(CONTENT_NAME);
+
 type SelectContentElement = SelectContentImplElement;
 interface SelectContentProps extends SelectContentImplProps {}
 
 const SelectContent = React.forwardRef<SelectContentElement, SelectContentProps>(
   (props: ScopedProps<SelectContentProps>, forwardedRef) => {
     const context = useSelectContext(CONTENT_NAME, props.__scopeSelect);
     const [fragment, setFragment] = React.useState<DocumentFragment>();
+    // We set this to true by default so that events bubble to forms without JS (SSR)
+    const isFormControl = context.trigger ? Boolean(context.trigger.closest('form')) : true;
+    const [nativeOptionsSet, setNativeOptionsSet] = React.useState(new Set<ReactOption>());
+
+    const nativeSelectKey = Array.from(nativeOptionsSet.values())
+      .map((option) => option.props.value)
+      .join(';');
 
     // setting the fragment in `useLayoutEffect` as `DocumentFragment` doesn't exist on the server
     useLayoutEffect(() => {
       setFragment(new DocumentFragment());
     }, []);
 
-    return context.open ? (
-      <SelectContentImpl {...props} ref={forwardedRef} />
-    ) : fragment ? (
-      ReactDOM.createPortal(
-        <SelectContentContextProvider scope={props.__scopeSelect}>
-          <Collection.Slot scope={props.__scopeSelect}>
-            <div>{props.children}</div>
-          </Collection.Slot>
-        </SelectContentContextProvider>,
-        fragment as any
-      )
-    ) : null;
+    return (
+      <SelectNativeOptionsProvider
+        scope={props.__scopeSelect}
+        onNativeOptionAdd={React.useCallback((option) => {
+          setNativeOptionsSet((prev) => new Set(prev).add(option));
+        }, [])}
+        onNativeOptionRemove={React.useCallback((option) => {
+          setNativeOptionsSet((prev) => {
+            const itemsSet = new Set(prev);
+            itemsSet.delete(option);
+            return itemsSet;
+          });
+        }, [])}
+      >
+        {isFormControl ? (
+          <BubbleSelect
+            key={nativeSelectKey}
+            aria-hidden
+            tabIndex={-1}
+            name={context.name}
+            autoComplete={context.autoComplete}
+            value={context.value}
+            // enable form autofill
+            onChange={(event) => context.onValueChange(event.target.value)}
+          >
+            {Array.from(nativeOptionsSet.values())}
```

Comment:
Just curious, is it required to explicitly access `values()`? or is `Array.from(nativeOptionsSet)`  equivalent?

Reply:
Oh good point, now that it's a set it's probably equivalent, I'll check.